### PR TITLE
config: change default initial fetch timeout to 15s

### PR DESCRIPTION
**Pull request by ramaraochavali**
Description: As discussed in PR https://github.com/envoyproxy/envoy/pull/7427, changing the default initial fetch time out to 15s.
Risk Level: Low
Testing: Changed
Docs Changes: Updated
Release Notes: Added

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@mattklein123  @htuch as discussed changed this. PTAL.

File: api/envoy/api/v2/core/config_source.proto [hunk 1]
```diff
@@ -112,13 +112,12 @@ message ConfigSource {
...
 
-  // Optional initialization timeout.
   // When this timeout is specified, Envoy will wait no longer than the specified time for first
   // config response on this xDS subscription during the :ref:`initialization process
   // <arch_overview_initialization>`. After reaching the timeout, Envoy will move to the next
   // initialization phase, even if the first config is not delivered yet. The timer is activated
   // when the xDS API subscription starts, and is disarmed on first config update or on error. 0
   // means no timeout - Envoy will wait indefinitely for the first xDS config (unless another
-  // timeout applies). Default 0.
+  // timeout applies). If not specified the default is 15000ms (15 seconds).
```
Comment (mattklein123) [inline api/envoy/api/v2/core/config_source.proto]: nit: I think you could just say "the default is 15s."

File: api/envoy/api/v2/core/config_source.proto [hunk 2]
```diff
@@ -112,13 +112,12 @@ message ConfigSource {
...
   google.protobuf.Duration initial_fetch_timeout = 4;
```
Comment (mattklein123) [inline api/envoy/api/v2/core/config_source.proto]: Should this have a constraint to make this non-zero if it is set?

File: docs/root/intro/version_history.rst [hunk 3]
```diff
@@ -3,6 +3,7 @@ Version history
 
 1.12.0 (pending)
 ================
+* config: changed the default value of :ref:`initial_fetch_timeout <envoy_api_field_core.ConfigSource.initial_fetch_timeout>` from 0s to 15s. This is a change in behaviour in the sense that Envoy initialization always completes within 15s by default. Refer to :ref:`initialization process <arch_overview_initialization>` for more details.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: I don't think it's true that Envoy will always complete init within 15s, given that other things have to happen like HC, etc. Can you rephrase?

File: api/envoy/api/v2/core/config_source.proto (hunk 2 above)
Comment (ramaraochavali) [reply api/envoy/api/v2/core/config_source.proto]: I was thinking about it, but decided against it because it may allow people get back old behaviour it they need. I was thinking of updating the doc with that info. WDYT?

Comment (ramaraochavali) [reply api/envoy/api/v2/core/config_source.proto]: doc already exists
  //  0 means no timeout - Envoy will wait indefinitely for the first xDS config (unless another
  // timeout applies)

Comment (mattklein123) [reply api/envoy/api/v2/core/config_source.proto]: OK sounds good.

**Comment by mattklein123** [comment]
> I think it's reasonable to set a default like this, but I wonder if this counts as a breaking API change; we're totally changing the behavior of Envoy initialization within a major API version?

I suppose it's a breaking change, but IMO it's for the better so FWIW I think it's OK. If we are concerned perhaps we should runtime guard it?

**Comment by ramaraochavali** [comment]
> . If we are concerned perhaps we should runtime guard it?

I agree with you that it is better change. Since the old behaviour is still supported with the same config, do we still want to guard with runtime? I am assuming runtime enables this by default.

**Comment by mattklein123** [comment]
My preference would be to enable this by default, the runtime would just store the default which would allow people to revert back to the previous behavior. Let's see what @htuch thinks.

**Comment by htuch** [comment]
Probably worth runtime guarding, since it will allow Envoy global reversion to the old behavior.

I'm thinking through how this relates to https://github.com/envoyproxy/envoy/issues/6271. Reading through the document, I think we decided that it doesn't matter what we set or change defaults for wrapped types to, since a management server that actually cares should set fields rather than relying on defaults. So, I think we are safe from a stable versioning perspective here.

**Comment by ramaraochavali** [comment]
>  So, I think we are safe from a stable versioning perspective here.
If we are safe from stable versioning perspective - I prefer not to introduce the runtime  guard as it introduces another level of config for this field. 

From operational perspective would runtime provide better way to revert - some how it needs to be set for all running envoys right?

**Comment by htuch** [comment]
We're good from an API stability perspective, so I'll leave this between you an @mattklein123

**Comment by ramaraochavali** [comment]
@htuch  Thank you.

@mattklein123  WDYT? should we runtime guard this given the default value we set here is more sensible?
